From 04b87b3ca37f1a9de987637f5b223c27c5aaa579 Mon Sep 17 00:00:00 2001
From: Sebastian Bauer <mail@sebastianbauer.info>
Date: Fri, 22 Jan 2016 20:04:50 +0100
Subject: [PATCH 08/29] Added libstc++ support for AmigaOS.

---
 .../os/{generic => amigaos}/ctype_base.h      |  2 +-
 .../{aix => amigaos}/ctype_configure_char.cc  | 24 +++++++++----------
 .../os/{generic => amigaos}/ctype_inline.h    | 16 ++++++-------
 .../os/{generic => amigaos}/error_constants.h |  6 ++---
 .../os/{vxworks => amigaos}/os_defines.h      | 13 +++++-----
 libstdc++-v3/configure.host                   |  3 +++
 6 files changed, 34 insertions(+), 30 deletions(-)
 copy libstdc++-v3/config/os/{generic => amigaos}/ctype_base.h (97%)
 copy libstdc++-v3/config/os/{aix => amigaos}/ctype_configure_char.cc (87%)
 copy libstdc++-v3/config/os/{generic => amigaos}/ctype_inline.h (96%)
 copy libstdc++-v3/config/os/{generic => amigaos}/error_constants.h (97%)
 copy libstdc++-v3/config/os/{vxworks => amigaos}/os_defines.h (86%)

diff --git a/libstdc++-v3/config/os/generic/ctype_base.h b/libstdc++-v3/config/os/amigaos/ctype_base.h
similarity index 97%
copy from libstdc++-v3/config/os/generic/ctype_base.h
copy to libstdc++-v3/config/os/amigaos/ctype_base.h
index e73721e57a5a82c47b9375fd8f939d9a798f4c56..260b6142b6fc6a11a1e5f0bf7820b592d09f285e 100644
--- a/libstdc++-v3/config/os/generic/ctype_base.h
+++ b/libstdc++-v3/config/os/amigaos/ctype_base.h
@@ -1,9 +1,9 @@
 // Locale support -*- C++ -*-
 
-// Copyright (C) 1997-2018 Free Software Foundation, Inc.
+// Copyright (C) 1997-2015 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
 // terms of the GNU General Public License as published by the
 // Free Software Foundation; either version 3, or (at your option)
 // any later version.
diff --git a/libstdc++-v3/config/os/aix/ctype_configure_char.cc b/libstdc++-v3/config/os/amigaos/ctype_configure_char.cc
similarity index 87%
copy from libstdc++-v3/config/os/aix/ctype_configure_char.cc
copy to libstdc++-v3/config/os/amigaos/ctype_configure_char.cc
index cb80c496e7157c356212a489a99807272f5b6728..e6f4895aee78da54bc6b1e01df00816206361c41 100644
--- a/libstdc++-v3/config/os/aix/ctype_configure_char.cc
+++ b/libstdc++-v3/config/os/amigaos/ctype_configure_char.cc
@@ -1,9 +1,9 @@
 // Locale support -*- C++ -*-
 
-// Copyright (C) 2011-2018 Free Software Foundation, Inc.
+// Copyright (C) 2011-2015 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
 // terms of the GNU General Public License as published by the
 // Free Software Foundation; either version 3, or (at your option)
 // any later version.
@@ -39,29 +39,29 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
 // Information as gleaned from /usr/include/ctype.h
 
   const ctype_base::mask*
   ctype<char>::classic_table() throw()
   { return 0; }
 
-  ctype<char>::ctype(__c_locale, const mask* __table, bool __del,
-		     size_t __refs)
-  : facet(__refs), _M_del(__table != 0 && __del),
-  _M_toupper(NULL), _M_tolower(NULL),
-  _M_table(__table ? __table : classic_table())
-  {
+  ctype<char>::ctype(__c_locale, const mask* __table, bool __del, 
+		     size_t __refs) 
+  : facet(__refs), _M_del(__table != 0 && __del), 
+  _M_toupper(NULL), _M_tolower(NULL), 
+  _M_table(__table ? __table : classic_table()) 
+  { 
     memset(_M_widen, 0, sizeof(_M_widen));
     _M_widen_ok = 0;
     memset(_M_narrow, 0, sizeof(_M_narrow));
     _M_narrow_ok = 0;
   }
 
-  ctype<char>::ctype(const mask* __table, bool __del, size_t __refs)
-  : facet(__refs), _M_del(__table != 0 && __del),
-  _M_toupper(NULL), _M_tolower(NULL),
+  ctype<char>::ctype(const mask* __table, bool __del, size_t __refs) 
+  : facet(__refs), _M_del(__table != 0 && __del), 
+  _M_toupper(NULL), _M_tolower(NULL), 
   _M_table(__table ? __table : classic_table())
-  {
+  { 
     memset(_M_widen, 0, sizeof(_M_widen));
     _M_widen_ok = 0;
     memset(_M_narrow, 0, sizeof(_M_narrow));
     _M_narrow_ok = 0;
   }
 
@@ -81,13 +81,13 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
   }
 
   char
   ctype<char>::do_tolower(char __c) const
   { return ::tolower((int) __c); }
 
-  const char*
+  const char* 
   ctype<char>::do_tolower(char* __low, const char* __high) const
   {
     while (__low < __high)
       {
 	*__low = ::tolower((int) *__low);
 	++__low;
diff --git a/libstdc++-v3/config/os/generic/ctype_inline.h b/libstdc++-v3/config/os/amigaos/ctype_inline.h
similarity index 96%
copy from libstdc++-v3/config/os/generic/ctype_inline.h
copy to libstdc++-v3/config/os/amigaos/ctype_inline.h
index 3a0ef4f9921c572f07dcfa56aae441fa2b15dba8..cfa0146ae6b70623c2fe63b864ceef7bce0d5fe0 100644
--- a/libstdc++-v3/config/os/generic/ctype_inline.h
+++ b/libstdc++-v3/config/os/amigaos/ctype_inline.h
@@ -1,9 +1,9 @@
 // Locale support -*- C++ -*-
 
-// Copyright (C) 2000-2018 Free Software Foundation, Inc.
+// Copyright (C) 2000-2015 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
 // terms of the GNU General Public License as published by the
 // Free Software Foundation; either version 3, or (at your option)
 // any later version.
@@ -27,16 +27,16 @@
  *  Do not attempt to use it directly. @headername{locale}
  */
 
 //
 // ISO C++ 14882: 22.1  Locales
 //
-
+  
 // ctype bits to be inlined go here. Non-inlinable (ie virtual do_*)
 // functions go in ctype.cc
-
+  
 // The following definitions are portable, but insanely slow. If one
 // cares at all about performance, then specialized ctype
 // functionality should be added for the native os in question: see
 // the config/os/bits/ctype_*.h files.
 
 // Constructing a synthetic "C" table should be seriously considered...
@@ -45,19 +45,19 @@ namespace std _GLIBCXX_VISIBILITY(default)
 {
 _GLIBCXX_BEGIN_NAMESPACE_VERSION
 
   bool
   ctype<char>::
   is(mask __m, char __c) const
-  {
+  { 
     if (_M_table)
       return _M_table[static_cast<unsigned char>(__c)] & __m;
     else
       {
 	bool __ret = false;
-	const size_t __bitmasksize = 15;
+	const size_t __bitmasksize = 15; 
 	size_t __bitcur = 0; // Lowest bitmask in ctype_base == 0
 	for (; __bitcur <= __bitmasksize; ++__bitcur)
 	  {
 	    const mask __bit = static_cast<mask>(1 << __bitcur);
 	    if (__m & __bit)
 	      {
@@ -109,29 +109,29 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
 		__ret |= __testis;
 	      }
 	  }
 	return __ret;
       }
   }
-
+   
   const char*
   ctype<char>::
   is(const char* __low, const char* __high, mask* __vec) const
   {
     if (_M_table)
       while (__low < __high)
 	*__vec++ = _M_table[static_cast<unsigned char>(*__low++)];
     else
       {
 	// Highest bitmask in ctype_base == 11.
-	const size_t __bitmasksize = 15;
+	const size_t __bitmasksize = 15; 
 	for (;__low < __high; ++__vec, ++__low)
 	  {
 	    mask __m = 0;
 	    // Lowest bitmask in ctype_base == 0
-	    size_t __i = 0;
+	    size_t __i = 0; 
 	    for (;__i <= __bitmasksize; ++__i)
 	      {
 		const mask __bit = static_cast<mask>(1 << __i);
 		if (this->is(__bit, *__low))
 		  __m |= __bit;
 	      }
diff --git a/libstdc++-v3/config/os/generic/error_constants.h b/libstdc++-v3/config/os/amigaos/error_constants.h
similarity index 97%
copy from libstdc++-v3/config/os/generic/error_constants.h
copy to libstdc++-v3/config/os/amigaos/error_constants.h
index af77f46be55ebc597d37b0dfbc2fd28ca803e3ac..74492cf61c875eac28cb8a86d5bfdb3b8807aa16 100644
--- a/libstdc++-v3/config/os/generic/error_constants.h
+++ b/libstdc++-v3/config/os/amigaos/error_constants.h
@@ -1,9 +1,9 @@
 // Specific definitions for generic platforms  -*- C++ -*-
 
-// Copyright (C) 2007-2018 Free Software Foundation, Inc.
+// Copyright (C) 2007-2015 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
 // terms of the GNU General Public License as published by the
 // Free Software Foundation; either version 3, or (at your option)
 // any later version.
@@ -90,16 +90,16 @@ _GLIBCXX_BEGIN_NAMESPACE_VERSION
       no_link = 				ENOLINK,
 #endif
 
       no_lock_available = 			ENOLCK,
 
 #ifdef _GLIBCXX_HAVE_ENODATA
-      no_message_available = 			ENODATA,
+      no_message_available = 			ENODATA, 
 #endif
 
-      no_message = 				ENOMSG,
+      no_message = 				ENOMSG, 
       no_protocol_option = 			ENOPROTOOPT,
       no_space_on_device = 			ENOSPC,
 
 #ifdef _GLIBCXX_HAVE_ENOSR
       no_stream_resources = 			ENOSR,
 #endif
diff --git a/libstdc++-v3/config/os/vxworks/os_defines.h b/libstdc++-v3/config/os/amigaos/os_defines.h
similarity index 86%
copy from libstdc++-v3/config/os/vxworks/os_defines.h
copy to libstdc++-v3/config/os/amigaos/os_defines.h
index 9f43cc4b9d5a2c52215722fec4ac1d895265e869..346f063958cd7e80ebf97be4acee0bdf391cb811 100644
--- a/libstdc++-v3/config/os/vxworks/os_defines.h
+++ b/libstdc++-v3/config/os/amigaos/os_defines.h
@@ -1,9 +1,9 @@
-// Specific definitions for VxWorks  -*- C++ -*-
+// Specific definitions for AmigaOS -*- C++ -*-
 
-// Copyright (C) 2000-2018 Free Software Foundation, Inc.
+// Copyright (C) 2000-2015 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
 // terms of the GNU General Public License as published by the
 // Free Software Foundation; either version 3, or (at your option)
 // any later version.
@@ -30,13 +30,14 @@
 #ifndef _GLIBCXX_OS_DEFINES
 #define _GLIBCXX_OS_DEFINES 1
 
 // System-specific #define, typedefs, corrections, etc, go here.  This
 // file will come before all others.
 
-//Keep vxWorks from defining min()/max() as macros
-#ifdef NOMINMAX
-#undef NOMINMAX
+// No ioctl() on AmigaOS
+#define _GLIBCXX_NO_IOCTL 1
+
+#ifdef __NEWLIB__
+#define _GLIBCXX_USE_C99_STDINT_TR1 1
 #endif
-#define NOMINMAX 1
 
 #endif
diff --git a/libstdc++-v3/configure.host b/libstdc++-v3/configure.host
index 155a3cdea1bad297b3cd4c21ecd3c7307b5e6ac9..9daab7c3c71169b10f5ff48c70e4d6d13228cf7f 100644
--- a/libstdc++-v3/configure.host
+++ b/libstdc++-v3/configure.host
@@ -219,12 +219,15 @@ case "${host_os}" in
     atomicity_dir="os/aix"
     ;;
   aix*)
     os_include_dir="os/generic"
     atomicity_dir="cpu/generic"
     ;;
+  amigaos*)
+    os_include_dir="os/amigaos"
+    ;;
   bsd*)
     # Plain BSD attempts to share FreeBSD files.
     os_include_dir="os/bsd/freebsd"
     ;;
   cygwin*)
     os_include_dir="os/newlib"
-- 
2.34.1

